libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,746 @@
+            vehicle.info.flags |= ADSB_FLAGS_VALID_CALLSIGN;
+        }
+
+        if (vehicle.info.flags != 0) {
+            frontend.handle_adsb_vehicle(vehicle);
+        }
+        break;
+
+    case MsgTypes_XP::TISB_StateVector_Report:
+    case MsgTypes_XP::TISB_ModeStatus_Report:
+    case MsgTypes_XP::TISB_CorasePos_Report:
+    case MsgTypes_XP::TISB_ADSB_Mgr_Report:
+        // TODO
+        return;
+
+    default:
+        return;
+    }
+
+}
+
+
+bool AP_ADSB_Sagetech::parse_byte_XP(const uint8_t data)
+{
+    switch (message_in_xp.state) {
+        default:
+        case ParseState::WaitingFor_Start:
+            if (data == 0xA5) {
+                message_in_xp.state = ParseState::WaitingFor_AssmAddr;
+            }
+            break;
+        case ParseState::WaitingFor_AssmAddr:
+            message_in_xp.state = (data == 0x01) ? ParseState::WaitingFor_MsgType : ParseState::WaitingFor_Start;
+            break;
+        case ParseState::WaitingFor_MsgType:
+            message_in_xp.packet.type = static_cast<MsgTypes_XP>(data);
+            message_in_xp.state = ParseState::WaitingFor_MsgId;
+            break;
+        case ParseState::WaitingFor_MsgId:
+            message_in_xp.packet.id = data;
+            message_in_xp.state = ParseState::WaitingFor_PayloadLen;
+            break;
+        case ParseState::WaitingFor_PayloadLen:
+            message_in_xp.packet.payload_length = data;
+            message_in_xp.index = 0;
+            message_in_xp.state = (data == 0) ? ParseState::WaitingFor_ChecksumFletcher : ParseState::WaitingFor_PayloadContents;
+            break;
+
+        case ParseState::WaitingFor_PayloadContents:
+            message_in_xp.packet.payload[message_in_xp.index++] = data;
+            if (message_in_xp.index >= message_in_xp.packet.payload_length) {
+                message_in_xp.state = ParseState::WaitingFor_ChecksumFletcher;
+                message_in_xp.index = 0;
+            }
+            break;
+
+        case ParseState::WaitingFor_ChecksumFletcher:
+            message_in_xp.packet.checksumFletcher = data;
+            message_in_xp.state = ParseState::WaitingFor_Checksum;
+            break;
+        case ParseState::WaitingFor_Checksum:
+            message_in_xp.packet.checksum = data;
+            message_in_xp.state = ParseState::WaitingFor_End;
+            if (checksum_XP(message_in_xp.packet)) {
+                parse_packet_XP(message_in_xp.packet);
+            }
+            break;
+
+        case ParseState::WaitingFor_End:
+            // we don't care if the end value matches
+            message_in_xp.state = ParseState::WaitingFor_Start;
+            break;
+    }
+    return false;
+}
+
+// compute Sum and FletcherSum and write them into msg.
+// returns true if the values in msg were already the correct ones
+//
+// this allows a single function to tell you if inbound msg is correct wirh true result, and ignore result for outbound

```

Comment:
This precludes having a `const` message hanging around which you want to check the checksum of.

Reply:
I really looks like these things are checksummed on the wire rather than having a checksum themselves.  i.e. it's wrong to be checksumming a `msg`.